### DISCUSSION TITLE
"Cannot specify both 'width' and 'min-width'"

### DISCUSSION THREAD

**Discussion by Enyium** [slint-ui/slint#6287] [General]
When having this code:

```slint
export component Demo {
    Rectangle {
        width: 100%;
        min-width: 50px;
    }
}
```

...I'm getting the error:

> Cannot specify both 'width' and 'min-width'

Is this reasonable? I can get a minimum applied, as if it would be on the outer `Rectangle` anyways, because it's carried forward from the inside, even with a non-percentage width (I know that it's `Demo` that the property gets finally effective on):

```slint
export component Demo {
    Rectangle {
        //width: 100%;
        width: 100px;

        background: gold;

        Rectangle {
            min-width: 50px;
        }
    }
}
```

Why is that rule in place? Shouldn't it be relaxed or given up?

**Comment by ogoffart** [comment]
I guess you can write something like this:
`width: parent.width.max(50px);`

You can't specify both width and min-width because width is also an alias to min-width and max-width.
If you could specify a width and it's smaller than min-width, then who wins? and what does rectangle.width evaluate to?

**Reply by Enyium** [reply]
`width: parent.width.max(50px);` could lead to overlapping rendering, couldn't it? It needs to propogate outwards. But with the usage site of an `@children` (i.e., on component usage), the parent inside the component it needs to propogate to isn't practicably accessible other than by letting the right constraint properties propogate by themselves. So, setting it on a parent can be not an option, and wrapping it in a parent for this purpose (at the usage site) would be a bad distribution of responsibilities.

`min-...` already wins over `max-...` (see #6204). And in this context, the properties evaluate to what's written, not their changed value (from a practical perspective). If this was to be changed, they both should also win over `width`/`height`.

But I understand that this could make one have reservations regarding the comprehensibility of bindings. Maybe, they *all* including `preferred-...` should evaluate to the practically applied value according to a precedence list (like programming language operator precedence). And then, of course, document it well. All in all, this seems like the most consistent to me.
